proposals/NNNN-dxil-vectors.md

```diff
@@ -0,0 +1,190 @@
+The returned vector value and the status indicator are grouped into a new `ResRet` helper structure type
+ that the load intrinsic returns.
+
+```asm
+  ; overloads: SM6.9: f16|f32|i16|i32
+  ; returns: status, vector
+  declare %dx.types.ResRet.v[NUM][TY] @dx.op.rawBufferVectorLoad.v[NUM][TY](
+      i32,                  ; opcode
+      %dx.types.Handle,     ; resource handle
+      i32,                  ; coordinate c0 (byteOffset)
+      i32,                  ; coordinate c1 (elementOffset)
+      i32)                  ; alignment
+```
+
+
+The return struct contains a single vector and a single integer representing mapped tile status.
+
+```asm
+  %dx.types.ResRet.v[NUM][TY] = type { vector<TYPE, NUM>, i32 }
+```
+
+Here and hereafter, `NUM` is the number of elements in the loaded vector, `TYPE` is the element type name,
+ and `TY` is the corresponding abbreviated type name (e.g. `i64`, `f32`).
+
+#### Vector access
+
+Dynamic access to vectors were previously converted to array accesses.
+Native vectors can be accessed using `extractelement`, `insertelement`, or `getelementptr` operations.
+Previously usage of `extractelement` and `insertelement` in DXIL didn't allow dynamic index parameters.
+
+#### Elementwise intrinsics
+
+A selection of elementwise intrinsics are given additional native vector forms.
+Elementwise intrinsics are those that perform their calculations irrespective of the location of the element
+ in the vector or matrix arguments except insofar as that position corresponds to those of the other elements
+ that might be used in the individual element calculations.
+An elementwise intrinsic `foo` that takes scalar or vector arguments could theoretically implement its vector version using a simple loop and the scalar intrinsic variant.
+
+```c++
+vector<TYPE, NUM> foo(vector<TYPE, NUM> a, vector<TYPE, NUM> b) {
+  vector<TYPE, NUM> ret;
+  for (int i = 0; i < NUM; i++)
+    ret[i] = foo(a[i], b[i]);
+}
+```
+  
+For example, `fma` is an elementwise intrinsic because it multiplies or adds each element of its argument vectors,
+ but `cross` is not because it performs an operation on the vectors as units,
+ pulling elements from different locations as the operation requires.
+
+The elementwise intrinsics that have native vector variants represent the
+ unary, binary, and tertiary generic operations:
+
+```asm
+ <[NUM] x [TYPE]> @dx.op.unary.v[NUM][TY](i32 opcode, <[NUM] x [TYPE]> operand1)
+ <[NUM] x [TYPE]> @dx.op.binary.v[NUM][[TY]](i32 opcode, <[NUM] x [TYPE]> operand1, <[NUM] x [TYPE]> operand2)
+ <[NUM] x [TYPE]> @dx.op.tertiary.v[NUM][TY](i32 opcode, <[NUM] x [TYPE]> operand1, <[NUM] x [TYPE]> operand2, <[NUM] x [TYPE]> operand3)
+```
+
+The scalarized variants of these DXIL intrinsics will remain unchanged and can be used in conjunction
+ with the vector variants.
+This means that the same language-level vector could be used in scalarized operations and native vector operations
```

Comment:
In what situations are we likely to see scalarization as opposed to vector results?

Reply:
Also, wanted to confirm that DXIL vectorization is not limited to HLSL *long* vectors but even applies to the original ones.

Reply:
The length of the native vectors isn't part of this spec, so native vectors can be any length. I tried to communicate that without calling out long vectors explicitly to avoid a codependency. I think if this spec were viewed outside the context of a PR where long vectors is included, that might be clearer.

As to what situations we will see scalarizations, if that's useful for a preview of this feature, we can definitely work that out what can be expected as we have no strong opinions except to enable a preview. At full release of this feature, the scalar versions are not going away and the vector versions will apply to a broad selection of opcodes. This means that DXIL consumers will ultimately be expected to handle vectorized or scalarized versions of any relevant operations.